compiler/rustc_middle/src/ty/sty.rs

```diff
@@ -1315,6 +1315,29 @@ impl<'tcx> Ty<'tcx> {
         cf.is_break()
     }
 
+    /// Checks whether a type recursively contains some `TyKind::Error(_)`.
+    ///
+    /// Example: The (erroneous) type `Vec<{type error}, {type error}>` returns
+    /// true.
+    pub fn contains_error(self) -> bool {
```

Comment:
How does this differ from [`.references_error()`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/trait.TypeVisitableExt.html#method.references_error)? I think this method is unnecessary.

Reply:
For sanity, let $R$ = `ty.references_error()` and $C$ = `ty.contains_error()`. $R$ checks if `ty` has `TypeFlags::HAS_ERROR`, while $C$ checks each contained type for being a `TyKind::Error`. 

I'm not knowledgeable enough about deeper usage of `TypeFlags` to clarify why, but $C$ seems to result in similar output to `R && ty.to_string().contains("{type error}")`. 

(I think `Ty`'s inner `GlobalCtxt` lets `FlagComputation::for_kind(...)` attribute errors to `ty` during `CtxtInterners::intern_ty(...)`, even if it's not about what `ty` directly contains.)

Anyways, when just using $R$, many tests fail as they're missing expected diagnostics. For example, `tests/ui/const-generics/issues/issue-56445-1.rs` fails with this:

```rust
--- not found errors (from test file) ---
ERROR     line   9: `&str` is forbidden as the type of a const generic parameter
```

It feels like $R$ is (loosely) a superset of $C$. Creating $C$ results in the appropriate output for this check.

Reply:
`tests/ui/const-generics/issues/issue-56445-1.rs` fails because it already has the preexisting error where `&str`'s lifetime is not resolvable, so the type ends up containing errors.

There is no reason to add a new method that is used nowhere else in the compiler -- please just use `references_error` and adjust the tests accordingly.

Reply:
I understand your concerns with adding new methods, but the diagnostics would plainly be worse with the existing method. 

For example, that test now complains about the lifetime, ignoring the type:

```rust
$ cargo +stage1 check
    Checking issue_118179 v0.1.0 (/home/barrett/Documents/projects/testing/rust_lang_repros/issue_118179)
error[E0770]: the type of const parameters must not depend on other generic parameters
  --> src/main.rs:10:26
   |
10 | struct Bug<'a, const S: &'a str>(PhantomData<&'a ()>);
   |                          ^^ the type must not depend on the parameter `'a`
   |
   = note: lifetime parameters may not be used in the type of const parameters

For more information about this error, try `rustc --explain E0770`.
error: could not compile `issue_118179` (bin "issue_118179") due to 1 previous error
```

A user, in order, would have to remove the lifetime, then fix the next error when it pops up, too. I don't like when that happens to me. :(

Now, compare that to the diagnostics with the new method:

```rust
$ cargo +stage1 check
    Checking issue_118179 v0.1.0 (/home/barrett/Documents/projects/testing/rust_lang_repros/issue_118179)
error[E0770]: the type of const parameters must not depend on other generic parameters
  --> src/main.rs:10:26
   |
10 | struct Bug<'a, const S: &'a str>(PhantomData<&'a ()>);
   |                          ^^ the type must not depend on the parameter `'a`
   |
   = note: lifetime parameters may not be used in the type of const parameters

error: `&str` is forbidden as the type of a const generic parameter
  --> src/main.rs:10:25
   |
10 | struct Bug<'a, const S: &'a str>(PhantomData<&'a ()>);
   |                         ^^^^^^^
   |
   = note: the only supported types are integers, `bool`, and `char`
help: add `#![feature(adt_const_params)]` to the crate attributes to enable more complex and user defined types
   |
8  + #![feature(adt_const_params)]
   |
help: add `#![feature(unsized_const_params)]` to the crate attributes to enable references to implement the `ConstParamTy` trait
   |
8  + #![feature(unsized_const_params)]
   |

For more information about this error, try `rustc --explain E0770`.
error: could not compile `issue_118179` (bin "issue_118179") due to 2 previous errors
```

Nonetheless, if you're comfortable with the loss of clarity, I can remove the new method. Please let me know what you think.

Reply:
Yeah, I don't care about fixing it in a single place. This is a more general issue with [`ReError`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/type.RegionKind.html#variant.ReError) that needs a delicate, global refactoring that I'd rather not tangle into this PR.

Reply:
That makes total sense - thank you for the link and explanation! I'll adjust the tests and use your suggestion.